### DISCUSSION TITLE
Dynamic model equations

### DISCUSSION THREAD

**Discussion by ThomasGiavasopoulos** [do-mpc/do-mpc#254] [Q&A]
Good afternoon!

Is it possible to include functions as part of the RHS of a state in a do-mpc model? 

For example, I have a function that calculates the moment of inertia of an object, given mass of fuel left in the rocket tanks. I currently include an expression for the change in mass of fuel in my do-mpc model.  Is there a way to directly include the fuel flow rate in the function calculating the instantaneous moment of inertia?

Thank you very much.

Thomas Giavasopoulos

**Comment by ThomasGiavasopoulos** [comment]
The mass of the rocket is a state of my system and passed as :

```
Rocket_mass = model.set_variable(var_type='_x', var_name='Rocket_mass', shape=(1,1))
model.set_rhs( -(Thrust) / (SpecificImpulse * AccelerationOfGravity) ) )   
``` 
I have a function that calculates Moment of Inertia only given the instantaneous mass. How should I express that variable in the dynamic model in do-mpc? I have tried passing the function as it is, but it fails since it contains non CasADi structures.

My goal is to take into account that changing moment of inertia as mass changes throughout the trajectory.

Thank you!

Thomas

**Comment by ThomasGiavasopoulos** [comment]
I have tried including the Moment of Inertia as a state, but this is not standard form, y_dot = f(y).

I now know that I have to calculate the moment of inertia from the current mass. As mentioned before, I have constructed a function that can do that, but I get an error if I pass it as a parameter since I am using np.arrays in the function. 

Is there any way to calculate and update the moment of inertia at each time step without it being included as a state?

Maybe this should be a changing parameter? 

Thank you very much for your time.

Thomas Giavasopoulos

**Comment by ThomasGiavasopoulos** [comment]
A time varying parameter is the feature that I am after. I have now defined the _tvp parameter as shown in the FAQ section of the do-mpc manual. Unfortunately, I am met with this error message:

<img width="715" alt="image" src="https://user-images.githubusercontent.com/57414585/161766855-befe1a64-0ed1-4822-8d38-f62ec2217e3d.png">

I am not sure why this message appears due to the fact that my code is identical to that of the FAQ section for the definition of the time-varying parameter. Any input is appreciated.

Thank you very much.

Thomas Giavasopoulos

**Comment by 4flixt** [comment]
Hi @ThomasGiavasopoulos,

all RHS expressions must be directly obtained from functions that operate only with CasADi objects that were introduced in the model (states, inputs, etc.). You cannot call external functions.

Regarding your problem with the ``tvp`` I cannot directly see the mistake.

Best
4flixt

**Reply by ThomasGiavasopoulos** [reply]
Thank you very much for your reply @4flixt !

I am currently working on adding the tvp expression appropriately, I will keep updating this discussion.

Thanks again.

Thomas

**Comment by ThomasGiavasopoulos** [comment]
Good afternoon,

I have been continuously trying to incorporate the moment of inertia variable discussed above, but I had no success. 

My model looks like this:
```
#Definition of states
x_position = model.set_variable(var_type='_x', var_name='x_position', shape=(1,1))
z_position = model.set_variable(var_type='_x', var_name='z_position', shape=(1,1))
theta_position = model.set_variable(var_type='_x', var_name='theta_position', shape=(1,1))
dv = model.set_variable(var_type='_x', var_name='dv', shape=(3,1))
Rocket_fuel_mass = model.set_variable(var_type='_x', var_name='Rocket_fuel_mass', shape=(1,1))

T_set = model.set_variable(var_type='_u', var_name='T_set')
T_angle_set = model.set_variable(var_type='_u', var_name='T_angle_set')
RightSmallThruster_set = model.set_variable(var_type='_u', var_name='RightSmallThruster_set')
LeftSmallThruster_set = model.set_variable(var_type='_u', var_name='LeftSmallThruster_set')
```

My intention is to recalculate Moment of Inertia at each time step. I do not require integration for its formula and it is not a state.I just wish to recalculate its value given the numbers found from the calculations of the states. I am assuming that this should look like this:
`### Rocket_MoI = model.set_variable(var_type='_tvp', var_name='Rocket_MoI', shape=(1,1))
`, but I unfortunately fail to use the tvp structure discussed in the documentation.

Is there any way that this can be done?

Thank you very much.

Thomas
